capture/src/partition_limits.rs

```diff
@@ -49,6 +50,23 @@ impl PartitionLimiter {
             gauge!("partition_limits_key_count", self.limiter.len() as f64);
         }
     }
+
+    /// Clean up the rate limiter state, once per minute. Ensure we don't use more memory than
+    /// necessary.
+    pub async fn clean_state(&self) {
+        // Give a small amount of randomness to the interval to ensure we don't have all replicas
+        // locking at the same time. The lock isn't going to be held for long, but this will reduce
+        // impact regardless.
+        let interval_secs = rand::thread_rng().gen_range(60..70);
+
+        let mut interval = tokio::time::interval(tokio::time::Duration::from_secs(interval_secs));
+        loop {
+            interval.tick().await;
+
+            self.limiter.retain_recent();
+            self.limiter.shrink_to_fit();
```

Comment:
nit: because we know that we'll see new distinct_ids soon (as sessions churn from visitor to visitor), I don't think the cost of `shrink` is worth it now, I'd just run `retain` every 5 minutes, which should in theory keep the active key count bounded. Alternatively, we could `shrink` before we `retain`, to only re-allocate the map if it didn't grow back after the last `retain`.
But it's safe to merge as is and tune later.